### PR TITLE
Abstract current OpenAI and AzureOpenAI implementations from the Core Kernel

### PR DESCRIPTION
**Pull request by RogerBarreto**
### Motivation and Context

This is an effort to abstract current OpenAI and AzureOpenAI implementations from the Core Kernel code base.

1. Why is this change required?
This change increases the abstractions on the Core and allow it to be handling any Backend implementations in the future

2. What problem does it solve?
Currently the Kernel is very coupled with those two implementations and any effort to do something differently demands more code and complexity

3. What scenario does it contribute to?
This contributes to reducing the places in the code that need to be changed to support new or any change in the current configurations

### Description

The changes made includes abstracting how backend configurations are added to the KernelConfig class making use of two new flag interfaces `ICompletionBackendConfig` and `IEmbeddingsBackendConfig`.

Another change that was necessary was inverting the control on the ITextCompletionClient instantiation in Kernel `CreateSemanticFunction` to the `KernelConfig` as a delegate when a  `func.SetAIBackend` is defined.

With those changes it was possible to move out of both `Kernel` and `KernelConfig` any specific implementations from AzureOpenAI and OpenAI and adding them as Extension methods in `KernelConfigExtensions.cs`.

This makes the code simpler and prepares a clear path if we need to move those implementations to a dedicated backend package and add many more implementations.

Added unit tests to validate the current behavior as expected without impacts.

### Contribution Checklist

- [x] The code builds clean without any errors or warnings
- [x] The PR follows SK Contribution Guidelines (https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md)
- [x] The code follows the .NET coding conventions (https://learn.microsoft.com/dotnet/csharp/fundamentals/coding-style/coding-conventions) verified with `dotnet format`
- [x] All unit tests pass, and I have added new tests where possible
- [x] I didn't break anyone :smile:

### CODE REVIEW

File: dotnet/src/SemanticKernel.Test/Configuration/KernelConfigTests.cs [hunk 1]
```diff
@@ -393,4 +396,158 @@ public void GetCompletionBackendItReturnsDefaultWhenNoLabelIsProvided()
...
+
+        ITextCompletionClient CreateFunc(ILogger logger) => fakeClient;
+        var fakeCompletion = new CompletionBackendConfigFake { Label = "test" };
+
+        // Act
+        target.AddCompletionBackendConfig(fakeCompletion, CreateFunc);
+        var createClientFunction = target.TryGetCompletionBackendCreateClient(fakeCompletion);
+        var createdClient = createClientFunction?.Invoke(mockLogger.Object);
+
+        Assert.Equal(fakeClient, createdClient);
```
[the remark this answers is not in the document]
Comment (RogerBarreto) [reply dotnet/src/SemanticKernel.Test/Configuration/KernelConfigTests.cs:515]: Fixed

File: dotnet/src/SemanticKernel/Configuration/ICompletionBackendConfig.cs [hunk 2]
```diff
@@ -0,0 +1,7 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+
+namespace Microsoft.SemanticKernel.Configuration;
+
+public interface ICompletionBackendConfig : IBackendConfig
```
[the remark this answers is not in the document]
Comment (RogerBarreto) [reply dotnet/src/SemanticKernel/Configuration/ICompletionBackendConfig.cs:8]: @dmytrostruk Comment added.

File: dotnet/src/SemanticKernel/AI/OpenAI/Configuration/KernelConfigExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+    /// <param name="apiKey">Azure OpenAI API key, see https://learn.microsoft.com/azure/cognitive-services/openai/quickstart</param>
+    /// <param name="apiVersion">Azure OpenAI API version, see https://learn.microsoft.com/azure/cognitive-services/openai/reference</param>
+    /// <param name="overwrite">Whether to overwrite an existing configuration if the same name exists</param>
+    /// <returns>Self instance</returns>
+    public static KernelConfig AddAzureOpenAICompletionBackend(this KernelConfig kernelConfig,
+        string label,
+        string deploymentName,
+        string endpoint,
+        string apiKey,
+        string apiVersion = "2022-12-01",
```
[the remark this answers is not in the document]
Comment (RogerBarreto) [reply dotnet/src/SemanticKernel/AI/OpenAI/Configuration/KernelConfigExtensions.cs]: Good suggestion, created a constant in the config file.

File: dotnet/src/SemanticKernel/AI/OpenAI/Configuration/KernelConfigExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+
+using Microsoft.SemanticKernel.AI.OpenAI.Services;
+using Microsoft.SemanticKernel.Configuration;
+
+namespace Microsoft.SemanticKernel.AI.OpenAI.Configuration;
+
+public static class KernelConfigExtensions
+{
+    /// <summary>
+    /// Adds an Azure OpenAI backend to the list.
```
[the remark this answers is not in the document]
Comment (RogerBarreto) [reply dotnet/src/SemanticKernel/AI/OpenAI/Configuration/KernelConfigExtensions.cs]: Good points, made the changes.

**Comment by dluc** [comment]
@RogerBarreto some conflicts to address. There's also another PR touching the HTTP layer, so could happen again, in one or the other, depending on the order these will be merged

**Comment by dluc** [comment]
100% agree with the premise and I've started drafting a doc to think through this holistically. 

Even if we merged this PR the problem would persist, e.g. the kernel is still rooted on text and completions (semantic memory is the only exception so far; and connectors are tangentially part of this conversation, e.g. OAI clients == connectors). 

We need to push further and reason about more generic and divers AI skills, and how the kernel allows to integrate them in the same programming model. 

My suggestion would be to park this PR, and join asap the conversation about the goals, and at the same time extend the list of models, making these changes part of the same story:
* decouple from completion
* integrate OAI ChatGPT, Edit, Insert, DallE, Stable Diffusion, Azure Speech
* allow to use any backend, and for each backend one must provide specs and clients

Adding more models is the only way to test if the new code can scale. Otherwise, I would say that we're just moving the code around, without verifying the benefits. (This would also be something like the 8th/9th time we refactor OpenAI clients integration, I think this is a good opportunity to extend the AI surface as planned).

**Comment by RogerBarreto** [comment]
> 100% agree with the premise and in fact I recently started drafting a doc to think through this holistically. I was trying to integrate a separate model and noticed the excessive effort required.
> 
> Even if we merged this PR the problem would persist, e.g. the kernel is still rooted on text and completions (semantic memory is the only exception so far; and connectors are tangentially part of this conversation, e.g. OAI clients == connectors).
> 
> We need to push further and reason about more generic and divers AI skills, and how the kernel allows to integrate them in the same programming model.
> 
> My suggestion would be to park this PR, and join asap the conversation about the goals, and at the same time extend the list of models, making these changes part of the same story:
> 
> * decouple from completion
> * integrate OAI ChatGPT, Edit, Insert, DallE, Stable Diffusion, Azure Speech
> * allow to use any backend, and for each backend one must provide specs and clients
> 
> Adding more models is the only way to test if the new code can scale. Otherwise, I would say that we're just moving the code around, without verifying the benefits. (This would also be something like the 8th/9th time we refactor OpenAI clients integration, I think this is a good opportunity to extend the AI surface as planned).

@dluc That make sense, we can leave this PR parked. 
Latests conflicts merged.

**Comment by RogerBarreto** [comment]
Closing for now as this approach will be revisited into a future PR

**Pull request closed without merging**